Update `join_keys` vignette

averissimo opened the pull request:
# Pull Request

Fixes #202

### Changes description

* Updates content
* Adds "Anatomy of `join_keys`" section
* Adds "Implicit relationships" section

vedhav commented on vignettes/join-keys.Rmd:
````diff
@@ -8,39 +8,98 @@ vignette: >
+jk
+```
+
+| Output of `print(jk)`              | Output annotation                    |
+| ---------------------------------- |:----------------------------------------:|
+| `## A join_keys object containing foreign keys between 3 datasets:` |  **Title** |
+| `## ds1: [id1]`                |  **Primary keys**<br />_\<dataset name\>_: [_\<primary_keys\>_]                                          | 
+| `##   <-- ds2: [id1]`          | **Foreign keys**<br /> _(arrow `<--` denotes `ds1` is the parent of `ds2`)_                    | 
+| `##   <-- ds3: [id1]`          |                                                                               | 
+| `## ds2: [id1, id2]`           |                                                                               | 
+| `##   --> ds1: [id1]`          | arrow `-->` denotes `ds2` is a child of `ds1`                                                                              | 
+| `##   --* (implicit via parent with): ds3` |  **Implicit relationship between `ds2` & `ds3`**<br />_(given that they share common keys with same parent)_    |
+| `## ds3: [id1, id3]`           |                                                                               | 
+| `##   --> ds1: [id1]`          |                                                                               | 
+| `##   --* (implicit via parent with): ds2` |                                                                               | 
+| `## ds4: [no primary keys]`    |                                                                               | 
+| `##   <-> ds5: [id5]`          |  **Foreign keys** <br />_(arrow `<->` denotes no parent definition between datasets)_      | 
+| `## ds5: [no primary keys]`    |                                                                               | 
+| `##   <-> ds4: [id4] `         |                                                                               | 
+
````
I love the idea of a table like this. The only improvement would be row borders so it's easy to read.
<img width="856" alt="Screenshot 2023-11-28 at 12 58 22 PM" src="https://github.com/insightsengineering/teal.data/assets/49812166/e2e7b2fe-5689-401c-990b-52f902d4988b">

averissimo replied:
I think the template being used on the site will add those rows

From [another vignete](https://insightsengineering.github.io/teal.data/latest-tag/articles/using-delayed-data-basic.html)

![image](https://github.com/insightsengineering/teal.data/assets/211358/bee46ad8-3413-4c75-8de3-b67c3825db25)

vedhav commented on vignettes/join-keys.Rmd:
```diff
@@ -8,39 +8,98 @@ vignette: >
+
+  ## Alternatively, when only a single primary key is defined the `join_keys()`
+  ##  wrapper can be skipped
+  # join_keys = join_key("IRIS", keys = "id")
```
```suggestion
  # join_keys = join_key("IRIS", keys = "id") # nolint
```

vedhav commented on vignettes/join-keys.Rmd:
```diff
@@ -49,20 +108,26 @@ data_1 <- data.frame(X = factor(1:10), Y = 21:30, Z = 1:10)
     join_key("D2", keys = c("V", "W"))
-    # join_key("D2", "D2", keys = c("V", "W")), # equivalent to using primary_key
+    ## equivalent to using primary_key
+    # join_key("D2", "D2", keys = c("V", "W")),
```
```suggestion
    # join_key("D2", "D2", keys = c("V", "W")), # nolint
```

averissimo replied:
Thanks, added a silent no lint to avoid having that code in the vignette

![image](https://github.com/insightsengineering/teal.data/assets/211358/7e4aad8c-4bf9-4875-b656-f68a870e6b67)

vedhav replied:
Even better!

vedhav reviewed (commented):
I think using `join_keys` object and switching to `teal_data` object is a little confusing. So, I feel that the complete vignette should use just manipulation using `join_keys`, even in the section `Primary key` and `Merge keys` and at the end there can be a section that talks about how it is relevant for a `teal_data` object and the same `join_key` methods can be applied directly on a `teal_data` object.

gogonzo commented on vignettes/join-keys.Rmd:
```diff
@@ -8,39 +8,98 @@ vignette: >
+- Merge keys, a concept similar to `SQL`'s foreign key.
 
-Usually, an application developer needs to specify the keys manually, but in case of datasets named according to the `ADaM` standard, `teal` can assign the keys on its own. 
+Usually, an application developer needs to specify the keys manually, but in case of datasets named according to the `ADaM` standard, `teal` can assign the keys on its own.
```
This is not true and vignette `including-adam-data-in-teal` shouldn't be mentioned. CDISC keys are handled in this package by relevant default (default_cdisc_join_keys and cdisc_data)

averissimo replied:
I was considering changing the documentation on that vignette, but that was wrong as the `default_cdisc_join_keys` lives here.

I've added a new section at the bottom. Please take a look at it.

gogonzo commented on vignettes/join-keys.Rmd:
````diff
@@ -8,39 +8,98 @@ vignette: >
   %\VignetteEncoding{UTF-8}
 ---
 
+```{r, include=FALSE}
+devtools::load_all()
+```
````
```suggestion
```

gogonzo commented on vignettes/join-keys.Rmd:
```diff
@@ -8,39 +8,98 @@ vignette: >
+  - `dataset_1`, `dataset_2` - name of each datasets (if `dataset_2` is the same as `dataset_1` then it creates a primary key)
+  - `key` - (optionally) named vector of column names
+
+Note that it is assumed that join keys are symmetric, i.e. `join_key("x", "y", "x_col" = "y_col")` will enable merge
+from "x" to "y" and vice versa.
```
```suggestion
Note that it is assumed that join keys are symmetric, i.e. `join_key("x", "y", c("x_col" = "y_col"))` will set relationship from "x" to "y" and vice versa.
```

averissimo commented:
@vedhav Do you think the "Anatomy of `join_keys`" section serves as good enough middle ground?

It talks about the functions and how to use them to create primary/foreign keys.

Maybe this section could be further improved :thinking: WDYT?

I think the following ones are nice for users as they show `join_keys` in a practical/realistic environment.

donyunardi commented:
Thanks for the work so far @averissimo!

What do you think about mentioning the benefits of `join_keys` in the context of `teal`? For example, by defining `join_keys`, it will assist us in creating a `merged_data` that can be used by module developers. 

In teal.transform, we have `merge_expression_module` function that will help users do this but maybe we can just mention this without going too deep.

kartikeyakirar commented:
@averissimo I was wondering if we could consider using italics or an alternative formatting approach instead of backticks for code element in headings . Specifically, when dealing with indexes or keep it as plain text . 
<img width="449" alt="image" src="https://github.com/insightsengineering/teal.data/assets/6700955/c000cad2-7e79-4280-b05c-b041eb900294">

 e.g something like this
<img width="459" alt="image" src="https://github.com/insightsengineering/teal.data/assets/6700955/8b442569-4fef-4336-a930-0e3f0b8d7506">

kartikeyakirar commented on vignettes/join-keys.Rmd:
```diff
@@ -10,37 +12,95 @@ vignette: >
+  join_keys = join_keys(join_key("IRIS", keys = "id"))
+
+  ## Alternatively, when only a single primary key is defined the `join_keys()`
+  ##  wrapper can be skipped
```
```suggestion
  ## wrapper can be skipped
```

averissimo replied:
I kinda like the second lines of comments to have the extra space, but that is arbitrary and personal.

Implemented it in 07982451. Thanks!

kartikeyakirar commented on vignettes/join-keys.Rmd:
```diff
@@ -49,51 +109,102 @@ data_1 <- data.frame(X = factor(1:10), Y = 21:30, Z = 1:10)
+
+### `ADaM` `join_keys`
+
+When working with datasets named according to the `ADaM` standard, `teal.data` can assign the keys on its own using the `teal.data::default_cdisc_join_keys` that contains primary and foreign keys for standard datasets (`r paste("\u0060", names(default_cdisc_join_keys), "\u0060", collapse = ", ", sep = "")`).
```
```suggestion
When working with datasets named according to the `ADaM` standard, `teal.data` can assign the keys on its own using the `teal.data::default_cdisc_join_keys` that contains primary and foreign keys for standard datasets as mentioned below.
```{r echo=FALSE}
names(default_cdisc_join_keys)
```
```

How about this for better visibility of datasets?

averissimo replied:
It looks great! thanks, done in 07982451

kartikeyakirar commented on vignettes/join-keys.Rmd:
```diff
@@ -49,51 +109,102 @@ data_1 <- data.frame(X = factor(1:10), Y = 21:30, Z = 1:10)
-within `teal` apps. For users familiar with `SQL` database schema, these relationships are symmetric and not as strict
-as `SQL` foreign key relationships as `teal` does not validate whether the values inserted into foreign key
-columns are present in the parent table.
+When passing multiple datasets to the `teal_data` function, dataset relationship are set using `join_keys` and `join_key` and these are used to merge datasets together within `teal` apps.
```
```suggestion
When passing multiple datasets to the `teal_data` function, dataset relationships are set using `join_keys` and `join_key` and these are used to merge datasets together within `teal` apps.
```

m7pr commented:
Hey @averissimo added feedback in this PR https://github.com/insightsengineering/teal.data/pull/209

averissimo commented:
@donyunardi great suggestion I added a small sub-section "Uses of join_keys in teal applications" _(I would welcome a better name though)_

Working on that I took my glasses off and tried to think like a new user and noticed that the vignette never addresses getters/setters.

I've added that after "Anatomy of Join Keys"

![image](https://github.com/insightsengineering/teal.data/assets/211358/fca19e64-1874-40af-861b-fc233823210d)

averissimo commented:
Thanks @m7pr

gogonzo commented on vignettes/join-keys.Rmd:
```diff
@@ -10,37 +12,138 @@ vignette: >
-ds <- dataset(
-  "IRIS",
-  iris_with_keys,
+td <- teal_data(
+  IRIS = iris_with_keys,
   code = "IRIS <- iris %>% mutate(id = factor(row_number()))",
-  keys = c("id")
+  join_keys = join_keys(join_key("IRIS", keys = "id"))
+
+  ## Alternatively, when only a single primary key is defined the `join_keys()`
+  ## wrapper can be skipped
+  # join_keys = join_key("IRIS", keys = "id")
 )
 
-ds$get_keys()
+join_keys(td)
```
Couple comments about this code chunk:
- iris changes names twice `iris -> iris_with_keys -> IRIS`
- code is reproducible but doesn't follow what happens actually with the object and some might be confused
- we should discourage writing `teal_data(..., code)` and promote `within` instead. 
- Alternatively, we can skip the code because example focus should be on `join_keys`

```suggestion
library(teal.data)
library(dplyr)

iris_df <- mutate(iris, id = factor(row_number()))

td <- teal_data(
  iris_df = iris_df,
  join_keys = join_keys(join_key("IRIS", keys = "id"))
)

join_keys(td)
```


The same applies to the next example. Please pick one way of specifying data in this vignette.

Either
```r
library(teal.data)
library(dplyr)

iris_df <- mutate(iris, id = factor(row_number()))

td <- teal_data(
  iris_df = iris_df,
  join_keys = join_keys(join_key("IRIS", keys = "id"))
)

join_keys(td)
```

or
```r
library(teal.data)
library(dplyr)

td <- within(
  teal_data(),
  iris_df <- mutate(iris, id = factor(row_number()))
)

join_keys(td) <- join_keys(join_key("IRIS", keys = "id"))

join_keys(td)
```

averissimo replied:
- Used `within` and added `datanames(...) <- c(...)` as the standard for the vignette.
  - It seems like promoting good practice

I've kept the examples using `teal_data` from the old vignette as it seemed like a useful and realistic way for the user to learn about `join_keys` _(the two are tightly related)_.

gogonzo reviewed (changes requested):
There are three different datanames sets (iris, D[1-3], ds[1-3). Please unify.

Also vignette first examples are using `join_keys` only which is fine, then `join_keys` are used with `teal_data` and again `join_keys` alone at the end. 
Let's start with explaining `join_keys` anatomy (only on `join_keys` object) and then use  `join_keys` with `teal_data` - and keep using the same the same datasets definition.

m7pr commented on vignettes/join-keys.Rmd:
````diff
@@ -10,37 +12,138 @@ vignette: >
+```{r, results="hide", message=FALSE, tidy=FALSE}
+library(teal.data)
+jk <- join_keys(
+  join_key("ds1", keys = "id1"),                  # ds1: [id1]
+  join_key("ds2", keys = c("id1", "id2")),        # ds2: [id1, id2]
+  join_key("ds3", keys = c("id1", "id3")),        # ds3: [id1, id3]
+  join_key("ds1", "ds2", keys = "id1"),           # ds1 <--> ds2
+  join_key("ds1", "ds3", keys = "id1"),           # ds1 <--> ds3
+  join_key("ds4", "ds5", keys = c("id4" = "id5")) # ds4 <--> ds5
````
unfortunately : (

averissimo replied:
I've added the `tidy=FALSE` on the chunk options that will ignore `styler` it for this chunk.

Having the comments aligned is informative, but I can remove that option and just eat up the extra spaces

gogonzo commented on vignettes/join-keys.Rmd:
```diff
@@ -10,90 +12,256 @@ vignette: >
+datanames(td_pk) <- c("ds1")
+
+join_keys(td_pk) <- join_keys(join_key("ds1", keys = "id"))
+## Alternatively, when only a single primary key is defined the `join_keys()`
+## the wrapper can be skipped
+# join_keys(td_pk) <- join_key("ds1", keys = "id")
```
I wouldn't necessarily write about this. Let's promote one way and don't give too much informations (alternatives).

```suggestion
```

averissimo commented on vignettes/join-keys.Rmd:
````diff
@@ -49,51 +109,102 @@ data_1 <- data.frame(X = factor(1:10), Y = 21:30, Z = 1:10)
+
+```{r}
+default_cdisc_join_keys
+```
````
I've changed the order, but didn't add `echo=FALSE` as that will remove any context of what the user is seeing.

However, I don't think that moving the full `default_cdisc_join_keys` to top will have any benefit as it will decrease the value of the other examples.

I won't die on this though :sweat_smile:

gogonzo replied:
Im fin with current state already. I had the biggest issue with data in the examples, it looks good now

chlebowa reviewed (commented):
The `--*` sign that means implicitly related datasets kind of looks like a modified `-->` but datasets that have `--*` will never be parent and child, more likely siblings.
How about something symmetric like `*--*`?

----
![image](https://github.com/insightsengineering/teal.data/assets/114988527/435d78d9-6e20-4095-9788-76bf1393ee52)

Please modify the `format` method either to say "dataset(s)" or add a condition on the number of datasets to print `dataset` or `datasets`. In `teal.slice` we use something like
`sprintf("dataset%s", if (n_datasets == 0L) "" else " ")`

chlebowa commented on vignettes/join-keys.Rmd:
```diff
@@ -10,90 +12,253 @@ vignette: >
-Usually, an application developer needs to specify the keys manually, but in case of datasets named according to the `ADaM` standard, `teal` can assign the keys on its own. 
-Refer to `vignette("including-adam-data-in-teal", package = "teal")` for more information.
+Typically, an application developer manually specifies these keys. However, for datasets following the `ADaM` standard, `teal.data` can automatically assign keys using the `default_cdisc_join_keys` object.
+Refer to the section ["`ADaM` `join_keys`"](#adam-join_keys) for details on using this object to select specific datasets.
```
I believe I tried it and it wasn't enough. It's been three days, though, I may remember wrong.

averissimo replied:
The link wasn't updated when section name was changed.

Corrected and "un-backticked" the section 5b04cdd9

gogonzo commented on vignettes/join-keys.Rmd:
````diff
@@ -10,90 +12,253 @@ vignette: >
+We can extend the previous example and define primary keys for multiple datasets:
 
 ```{r, message=FALSE}
 library(teal.data)
````
I agree with @averissimo  it is nicer when each code-chunk has own library calls so they are "independent".

averissimo replied:
I went back and forth a lot with this one, both with the library and the object raw build on every chunk (vs. progressive manipulation of an object).

From an outside perspective I think that this indendency works better.

chlebowa replied:
I'm fine with either as long as it's applied consistently. If packages is attached in every block, objects should start from scratch as well.

Just a note, I believe it is customary to attach the package only once.

averissimo commented on inst/WORDLIST:
```diff
@@ -1,3 +1,4 @@
+ADaM
 CDISC
```
:+1: I should've looked in other packages

Glancing at `{teal.slices}` it actually looks good on the pkgdown site. It may be just local vignettes that look weird.

averissimo commented on vignettes/join-keys.Rmd:
```diff
@@ -10,90 +12,253 @@ vignette: >
+The primary function of `join_keys` in `teal` applications is to facilitate the seamless [merging of datasets](https://insightsengineering.github.io/teal.transform/latest-tag/articles/data-merge.html) using `teal.transform`.
+
+Additionally, it plays a role on the data filtering using the _[Filter Panel](https://insightsengineering.github.io/teal/main/articles/filter-panel.html)_ in a `teal` application.
+The filtering of a _(parent)_ dataset triggers additional background actions that will look for all datasets in which foreign keys target that _parent_ and then filter those accordingly.
```
Yup

Done

```diff
diff --git a/vignettes/join-keys.Rmd b/vignettes/join-keys.Rmd
index d97117ef..a3e937b5 100644
--- a/vignettes/join-keys.Rmd
+++ b/vignettes/join-keys.Rmd
@@ -27,7 +27,7 @@ Refer to the section ["Joining Keys with `ADaM` Datasets"](#join-keys-in-adam-da
 The primary function of `join_keys` in `teal` applications is to facilitate the seamless [merging of datasets](https://insightsengineering.github.io/teal.transform/latest-tag/articles/data-merge.html) using `teal.transform`.
 
 Additionally, it plays a role on the data filtering using the _[Filter Panel](https://insightsengineering.github.io/teal/main/articles/filter-panel.html)_ in a `teal` application.
-The filtering of a _(parent)_ dataset triggers additional background actions that will look for all datasets in which foreign keys target that _parent_ and then filter those accordingly.
+The filters applied to a _(parent)_ dataset are also applied to their children.
 
 ## Anatomy of Join Keys
```

gogonzo commented on vignettes/join-keys.Rmd:
```diff
@@ -10,90 +12,253 @@ vignette: >
+
+## Accessing and Modifying keys
+
+The `join_keys` object behaves like a list and the preferred method to get a specific key to use the _subset_ operator with 2 indices. 
```
@averissimo this is still not addressed

averissimo replied:
Done in 598e66af

gogonzo commented on vignettes/join-keys.Rmd:
````diff
@@ -10,90 +12,253 @@ vignette: >
+```
+
+The full set of join keys can be used, or a subset of those.
+Please note, that if the subsetted dataset has a parent, that will also be returned.
````
@averissimo

averissimo replied:
Done in e0982d7f

averissimo commented on vignettes/join-keys.Rmd:
```diff
@@ -10,90 +12,253 @@ vignette: >
+Additionally, it plays a role on the data filtering using the _[Filter Panel](https://insightsengineering.github.io/teal/main/articles/filter-panel.html)_ in a `teal` application.
+The filtering of a _(parent)_ dataset triggers additional background actions that will look for all datasets in which foreign keys target that _parent_ and then filter those accordingly.
+
+## Anatomy of Join Keys
```
Proposal:

![image](https://github.com/insightsengineering/teal.data/assets/211358/cefd3e5f-5c10-4fca-80ae-9fe9e898755b)

chlebowa replied:
These are comments, not specific editing requests:

* _by `dataname` reference_ is not clear.

* A new `join_keys` can also be created empty or with only a primary key.

* The `join_keys` function is used as a constructor as well as to get or set join keys in `teal_data` objects.

* Returning itself is unclear. The function returns itself?

* _Is_ a joining key "represented by a pair of datasets and a character vector"? Or is a joining key a (named) character vector that we know to be relevant to a pair of datasets? The way I always think about it is a joining key is a column mapping between two datasets. Maybe we can improve this phrasing a little?

averissimo commented on vignettes/join-keys.Rmd:
```diff
@@ -10,90 +12,253 @@ vignette: >
-  iris_with_keys,
-  code = "IRIS <- iris %>% mutate(id = factor(row_number()))",
-  keys = c("id")
+td_pk <- within(
```
Done in fecf32f4

averissimo commented on vignettes/join-keys.Rmd:
````diff
@@ -10,90 +12,253 @@ vignette: >
+jk["ds5", "ds4"]
+```
+
+Modifying or adding a key uses the same notation with the assignment operator `<-`. This will ensure symmetry as it will create both relationships. Assigning `NULL` value will delete the relation.
````
Done in 98d0be50

averissimo commented on vignettes/join-keys.Rmd:
```diff
@@ -10,90 +12,253 @@ vignette: >
+### Implicit Relationships
+
+This is a special inferred relationship from existing `join_keys` that does not need to be explicitly defined, however, as any other it can be overwritten.
+If two datasets share common _foreign_ keys to the same _parent_ dataset, then there is an implicit relationship between them which is seamlessly accessible via the `join_keys` getter.
```
Done in d2cb4c95
